Add spi-empty-preamble-frames to WS2812 SPI driver

matt-wood-ct opened the pull request:
Introduced spi-empty-preamble-frames to handle SPI masters that don't drive bus lines low when idle. Updated driver to account for preamble frames, ensuring a clean bus reset before data transmission.

Designed to address #75965 and older issue #66664

matt-wood-ct commented:
I've checked the memory impact of this change, if you leave the feature unused the RAM usage is zero.
ROM usage change is +32 bytes (driver total ROM size is 431 bytes).

When you do use the feature the RAM impact is 1 byte per byte of preamble as expected.

Before changes:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       45304 B       512 KB      8.64%
             RAM:        8640 B        64 KB     13.18%
        IDT_LIST:          0 GB        32 KB      0.00%
```
After merge without preamble usage:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       45336 B       512 KB      8.65%
             RAM:        8640 B        64 KB     13.18%
        IDT_LIST:          0 GB        32 KB      0.00%
```
With 64 byte preamble:
```
Memory region         Used Size  Region Size  %age Used
           FLASH:       45336 B       512 KB      8.65%
             RAM:        8704 B        64 KB     13.28%
        IDT_LIST:          0 GB        32 KB      0.00%
```

All report above generated with `nrf52dk/nrf52832` used as the target

thedjnK commented:
Since this is completely useless for nearly all boards, I would prefer it if the spi driver configuration for your board could be fixed instead, and if not, using macros here so that the overhead for people not using this feature is 0 bytes

matt-wood-ct commented:
> Since this is completely useless for nearly all boards, I would prefer it if the spi driver configuration for your board could be fixed instead, and if not, using macros here so that the overhead for people not using this feature is 0 bytes

@thedjnK Understood, would it be acceptable for me to add a kconfig to remove all the new code and default that config to disabled, so I can keep the binding for DT config but have it unused unless the conf is enabled?

simonguinot commented:
@thedjnK @soburi @IVandeVeire @bjarki-andreasen what do you think of this PR ? It is a little bit like https://github.com/zephyrproject-rtos/zephyr/pull/66664. I am still not a big fan of this preamble thing but perhaps it is an acceptable compromise.

simonguinot commented on dts/bindings/led_strip/worldsemi,ws2812-spi.yaml:
```diff
@@ -33,3 +38,8 @@ properties:
+  spi-empty-preamble-frames:
+    type: int
+    default: 0
+    description: Number of empty frames to send before the data burst.
```
Since both the SPI frequency and the reset delay are known from the driver, then the preamble frame size can be computed without any other input.

matt-wood-ct replied:
That is a good point, let me rework this, just part way through adding the kconfig to toggle the whole thing off.

simonguinot commented on drivers/led_strip/ws2812_spi.c:
```diff
@@ -45,6 +45,7 @@ struct ws2812_spi_cfg {
 	const uint8_t *color_mapping;
 	size_t length;
 	uint16_t reset_delay;
+	uint16_t empty_preamble_frames;
```
And then it doesn't need to be stored.

matt-wood-ct replied:
I am not sure how to avoid storing this because in principle it can vary from instance to instance so it cannot be easily referenced in `ws2812_strip_update_rgb`

simonguinot replied:
Well, it is a problem.

thedjnK commented:
Can't the nxp driver actually work properly without this though if the configuration is changed? @dleach02 ?

matt-wood-ct commented:
@simonguinot Rework complete, there is now a kconfig for toggling the feature on/off and the number of preamble frames is calculated from `spi-max-frequency` and `reset-delay`

IVandeVeire commented:
@simonguinot This seems to be the same kind of fix as #66311 which worked for me. So I assume this will work too. Currently I am no longer working on the project where that issue occurred.

matt-wood-ct commented:
> @simonguinot This seems to be the same kind of fix as #66311 which worked for me. So I assume this will work too. Currently I am no longer working on the project where that issue occurred.

@IVandeVeire Unsurprisingly that issue was relating to an I.MX RT platform too.

matt-wood-ct commented:
@bjarki-andreasen Can you please help me understand how to address the compliance failures, I can see several issues I do not understand, namely:
It seems to have a problem with my signed-off-by entries despite them looking OK to me, the email matches my github account.
It says there is trailing whitespace in the binding YAML files I modified but I cannot see any on the lines mentioned and the file is not modified in my PR as the changes were removed during rework based on review feedback.

I can see that the complaints are on commit hashs I do not recognise and do not appear in my PR commit list?

- c12228a134aa08bdae948a8007d31ca302211b83
- 3d8cf34d02864fd44843a2f235f489a522ea586a

I think these commits were may have been destroyed when I force pushed some commit reworks, is that not the correct way to handle commit rework?
Sorry if these question are silly, I'm new to contributing, still figuring it out 😄

simonguinot reviewed (commented):
@matt-wood-ct please reorganize/merge the commits. It will help for the review.

matt-wood-ct committed:
drivers: led_strip: ws2812_spi: Add SPI preamble support

This commit introduces a configuration option to enable SPI preamble
support for WS2812 LED strip drivers. The preamble allows for proper
MOSI high behavior on some platforms, like the I.MX RT series MCUs.
Conditional code blocks and macros are added to handle this new
configuration seamlessly.

Signed-off-by: Matt Wood <matt.wood@corintech.com>

matt-wood-ct commented:
> @matt-wood-ct please reorganize/merge the commits. It will help for the review.

@simonguinot Ok, I have done so, hopefully also integrating the advice from @bjarki-andreasen .
Thanks both 😄

matt-wood-ct commented:
After some testing on hardware this morning, I've found on the I.MX RT 1040 the preamble is not enough to solve the issue, as I did in my original demo it seems you need a clean preamble and postamble. I will need to rework my changes to add a kconfig for that as an option too.
I tested using 85uS preamble and I found that he first 2 LEDs in the chain display the same colour despite the packets decoding acceptably in my logic analyser.
I also tested with just a 85uS postamble (no preamble) and that yielded strange behaviour, incorrect colours showing and LEDs combining / being skipped.

But naturally when I went back to the same setup I used in my original issue thread, a full reset low period on the head and tail of the burst, the behaviour was consistent and correct. So that seems to be test best implementation for using WS2812B LEDs on the NXP I.MX RT platform.

simonguinot commented on drivers/led_strip/Kconfig.ws2812:
```diff
@@ -17,6 +17,14 @@ config WS2812_STRIP_SPI
 
+config WS2812_STRIP_SPI_PREAMBLE_SUPPORT
+	bool "WS2812 LED strip SPI preamble support"
+	default n
```
Ideally, the default value should be set based on whether a device tree SPI attribute is present (like `spi-idle-high` for example). This SPI behavior should be described somehow in device tree.

thedjnK replied:
It would need review from SPI maintainers but don't really see why it should need more than that

thedjnK replied:
@tbursztyka @teburd would you be ok with introduction of such a flag?

matt-wood-ct replied:
Ok, as a compromise I have added this line
`default y if SOC_FAMILY_NXP_IMXRT`
Which will automatically enable this config if the driver is used on a IMXRT family SOC, its not a perfect solution, but it is functional and non intrusive.

simonguinot commented on drivers/led_strip/ws2812_spi.c:
```diff
@@ -83,16 +86,25 @@ static int ws2812_strip_update_rgb(const struct device *dev,
 	struct spi_buf buf = {
 		.buf = cfg->px_buf,
+#if IS_ENABLED(CONFIG_WS2812_STRIP_SPI_PREAMBLE_SUPPORT)
+		.len = cfg->empty_preamble_frames + (cfg->length * 8 * cfg->num_colors),
```
This driver follows the old 80 column coding style (except in one place which should be fixed). Please let's not mix styles and make this patch stick to a line length of 80 columns.

matt-wood-ct replied:
Agree, however this line has been reverted in the addition commit for postambles

matt-wood-ct committed:
drivers: led_strip: ws2812_spi: Add SPI postamble support

Adds a kconfig to enable a postamble of bus low time to reset
the LEDs between bursts, this ensures sensible behaviour on
NXP i.MX RT platforms.

Signed-off-by: Matt Wood <matt.wood@corintech.com>

simonguinot commented on drivers/led_strip/ws2812_spi.c:
```diff
@@ -182,8 +194,18 @@ static const struct led_strip_driver_api ws2812_spi_api = {
 #define WS2812_SPI_ZERO_FRAME(idx) \
 	(DT_INST_PROP(idx, spi_zero_frame))
+/* Get the latch/reset delay from the "reset-delay" DT property. */
+#define WS2812_RESET_DELAY(idx) \
```
The new line above is not needed. Please keep this line as is.

matt-wood-ct replied:
Agree, adjusted

simonguinot commented on drivers/led_strip/ws2812_spi.c:
```diff
@@ -182,8 +194,18 @@ static const struct led_strip_driver_api ws2812_spi_api = {
+	((DT_INST_PROP(idx, spi_max_frequency) * WS2812_RESET_DELAY(idx))/8000000)
+#else
+#define WS2812_PREAMBLE_FRAMES(idx) \
+	(0)
```
It can all fit on a single line.

matt-wood-ct replied:
Agree, adjusted

simonguinot reviewed (changes requested):
Thanks for the patch @matt-wood-ct but this reinforces my initial opinion: I don't like this work-around.

IMO with this patch we are butchering generic driver code to work around an issue we only have on few platforms. And even on this platforms there is probably better to do than just sending a bunch of zeros to keep the MOSI line low. For example on IMXRT1040 SoC, looking at your logic analyzer captures (#75965), we can see that the signal stays low after transmission. I think it is something that should be explored.

On the patch itself, I only have minor comments. Please see them below.

Let's wait for others to comment.

matt-wood-ct committed:
drivers: led_strip: ws2812_spi: Feedback corrections

Line length and whitespace adjustments based on review feedback.

Signed-off-by: Matt Wood <matt.wood@corintech.com>

thedjnK reviewed (commented):
I re-affirm my comment that I do not like this, if there is an issue with SPI then NXP should look into it and comment, and if something needs to be fixed, it should be fixed at SPI level, not a single driver that uses any supported SPI instance

matt-wood-ct commented:
> I re-affirm my comment that I do not like this, if there is an issue with SPI then NXP should look into it and comment, and if something needs to be fixed, it should be fixed at SPI level, not a single driver that uses any supported SPI instance

I don't disagree, but from a practicality point of view, is anyone on the NXP side going to do that? And is this idle bus behaviour a problem for any SPI slave implementation or is it just a quirk of using the SPI to emulate the WS2812B single wire comms?

If we are saying this is only a problem when using the SPIM for driving WS2812B then the solution should be in the WS2812B driver IMO

simonguinot reviewed (changes requested):
Hi @matt-wood-ct,

Thanks for your work on this issue.

But I am still not convinced. Like @thedjnK, I don't like this workaround. It really looks like a heavy hack. It seems wrong to fill a buffer with zeros before and after sending data just to ensure that the MOSI line is low.

It is true that this is not an SPI bug because the data acquisition is done on the clock signal. So the MOSI signal at idle state is not a problem in the case of "real" SPI communication.

But I still think that the fix must be done at the SPI or board level (pinctrl).

bjarki-andreasen commented:
> Hi @matt-wood-ct,
> 
> Thanks for your work on this issue.
> 
> But I am still not convinced. Like @thedjnK, I don't like this workaround. It really looks like a heavy hack. It seems wrong to fill a buffer with zeros before and after sending data just to ensure that the MOSI line is low.
> 
> It is true that this is not an SPI bug because the data acquisition is done on the clock signal. So the MOSI signal at idle state is not a problem in the case of "real" SPI communication.
> 
> But I still think that the fix must be done at the SPI or board level (pinctrl).

I would agree, if we where not already hacking by using a SPI controller as a protocol engine for a specific LED driver protocol it was not designed for. Using SPI here at all is a "hack". If the protocol requires MOSI be kept low, and the way to ensure that is to send out 0's. then that's what we should do.

matt-wood-ct committed:
Merge branch 'zephyrproject-rtos:main' into 75965_ws2812_spi_mosi_high_idle_support

matt-wood-ct committed:
drivers: led_strip: ws2812_spi: Configure based on SOC

Signed-off-by: Matt Wood <matt.wood@corintech.com>

matt-wood-ct committed:
drivers: led_strip: ws2812_spi: fixup

Signed-off-by: Matt Wood <matt.wood@corintech.com>

matt-wood-ct commented on drivers/led_strip/ws2812_spi.c:
```diff
@@ -45,7 +45,8 @@ struct ws2812_spi_cfg {
 	const uint8_t *color_mapping;
```
Understood, I will tidy the commits once the dialogs have closed off

matt-wood-ct replied:
Also I know I need to adjust it to a rebase, again I will sort this once everyone is happy with the general concepts & implementation

matt-wood-ct commented on drivers/led_strip/ws2812_spi.c:
```diff
@@ -86,24 +86,27 @@ static int ws2812_strip_update_rgb(const struct device *dev,
 #if IS_ENABLED(CONFIG_WS2812_STRIP_SPI_PREAMBLE_SUPPORT)
-	uint8_t *px_buf = &cfg->px_buf[cfg->empty_preamble_frames];
+    buf.len += cfg->empty_padding_frames;
+	uint8_t *px_buf = &cfg->px_buf[cfg->empty_padding_frames];
```
Understood and agree, reshuffled statements so the code which comes out always has the declaration first

matt-wood-ct commented on drivers/led_strip/ws2812_spi.c:
```diff
@@ -86,24 +86,27 @@ static int ws2812_strip_update_rgb(const struct device *dev,
+#endif
+#if IS_ENABLED(CONFIG_WS2812_STRIP_SPI_PREAMBLE_SUPPORT) || IS_ENABLED(CONFIG_WS2812_STRIP_SPI_POSTAMBLE_SUPPORT)
+    /* If zero values are padded on either end we need to ensure the buffer is all zeros before filling */
+	memset(cfg->px_buf, 0, buf.len);
```
As it stands it appears the buffer is not initialised, so it could and likely does have non zero values before this statement. That being said, it might be better to zero init the buffer during declaration on line 235

matt-wood-ct commented on drivers/led_strip/ws2812_spi.c:
```diff
@@ -197,15 +200,25 @@ static const struct led_strip_driver_api ws2812_spi_api = {
-	(WS2812_PREAMBLE_FRAMES(idx) + (WS2812_NUM_COLORS(idx) * 8 * WS2812_SPI_NUM_PIXELS(idx)))
+	((WS2812_PADDING_FRAMES(idx) \
+    * (IS_ENABLED(CONFIG_WS2812_STRIP_SPI_PREAMBLE_SUPPORT) + IS_ENABLED(CONFIG_WS2812_STRIP_SPI_POSTAMBLE_SUPPORT))) \
+    + (WS2812_NUM_COLORS(idx) * 8 * WS2812_SPI_NUM_PIXELS(idx)))
```
So the total extra ram consumption for this feature is:
2 bytes for holding the calculated pad length
(SPI_FREQUENCY * RESET_DELAY / 8,000,000) Bytes for each amble (so max 2x)
In practice the total additional RAM usage for a typical IMXRT setup is 77 bytes. 
IMO this level of memory overhead for a processor typically configured with MB scale memories is more then reasonable. As it stand the IMXRT platform is the only one with this issue.

As a general statement I would say this feature uses too much memory, but unfortunately this is the only practical way I can see the WS2812B SPI driver operating on NXPs IMXRTxxxx LPSPI system without wider butchering. 
This will remain completely compiled out when being used on normal behaviour SPI platforms.

matt-wood-ct commented on drivers/led_strip/ws2812_spi.c:
```diff
@@ -197,15 +200,25 @@ static const struct led_strip_driver_api ws2812_spi_api = {
-#define WS2812_PREAMBLE_FRAMES(idx) \
+#if IS_ENABLED(CONFIG_WS2812_STRIP_SPI_PREAMBLE_SUPPORT) || IS_ENABLED(CONFIG_WS2812_STRIP_SPI_POSTAMBLE_SUPPORT)
+#define WS2812_PADDING_FRAMES(idx) \
+	((DT_INST_PROP(idx, spi_max_frequency) * WS2812_RESET_DELAY(idx))/8000000)
```
Fixed up

matt-wood-ct commented on drivers/led_strip/ws2812_spi.c:
```diff
@@ -182,8 +194,18 @@ static const struct led_strip_driver_api ws2812_spi_api = {
+	(DT_INST_PROP(idx, reset_delay))
+#if IS_ENABLED(CONFIG_WS2812_STRIP_SPI_PREAMBLE_SUPPORT)
+#define WS2812_PREAMBLE_FRAMES(idx) \
+	((DT_INST_PROP(idx, spi_max_frequency) * WS2812_RESET_DELAY(idx))/8000000)
```
Fixed up

matt-wood-ct commented on drivers/led_strip/ws2812_spi.c:
```diff
@@ -100,12 +101,13 @@ static int ws2812_strip_update_rgb(const struct device *dev,
-    /* If zero values are padded on either end we need to ensure the buffer is all zeros before filling */
+#if IS_ENABLED(CONFIG_WS2812_STRIP_SPI_PREAMBLE_SUPPORT) || \
+	IS_ENABLED(CONFIG_WS2812_STRIP_SPI_POSTAMBLE_SUPPORT)
+	/* If zero values are padded on either end we need to ensure the buffer is
```
Fixed up

matt-wood-ct commented on drivers/led_strip/ws2812_spi.c:
```diff
@@ -86,24 +86,27 @@ static int ws2812_strip_update_rgb(const struct device *dev,
 	size_t i;
 	int rc;
+#if IS_ENABLED(CONFIG_WS2812_STRIP_SPI_POSTAMBLE_SUPPORT)
+    buf.len += cfg->empty_padding_frames;
```
Fixed up
